files/usr/share/jeos-firstboot/jeos-firstboot-functions

```diff
@@ -0,0 +1,155 @@
+#!/bin/bash
+stty_size() {
+	set -- `stty size`; LINES=$1; COLUMNS=$2
+	# stty size can return zero when not ready or
+	# its a serial console
+	if [ "$COLUMNS" = "0" -o "$LINES" = "0" ]; then
+		LINES=24
+		COLUMNS=80
+	fi
+}
+stty_size
+
+result=
+list=
+password=''
+modules=()
+
+let dh_menu=LINES-15
+let dh_text=LINES-5
+
+if pushd "/usr/share/jeos-firstboot/modules" &>/dev/null; then
+        for module in *; do
+                if [ -f "${module}" ] && source "${module}"; then
+                        modules+=("${module}")
+                fi
+        done
+        popd &>/dev/null
+fi
+
+call_module() {
+	local module="$1"
+	local module_func="$2"
+	module_function="${module}_${module_func}"
+	[ "$(type -t "${module_function}")" = "function" ] || return 1
+	"${module_function}" && true # To not trigger errexit
+	ret=$?
+	[ $ret -eq 0 ] || return $ret
+}
+
+call_module_hook() {
+        local hook="$1"
+        for module in "${modules[@]}"; do
+		call_module ${module} ${hook} || continue
```

Comment:
To restore the same behaviour as before:

```suggestion
		call_module ${module} ${hook} || return $?
```

Reply:
the "return" will change the behaviour, the current behaviour is actually ignoring errors, for example, the rapsberrywifi does not have a cleanup or post hook, and will fail. I'm thinking in including later one code to guarantee the integrity of the modules itself, check which functions and hooks they really support.

Reply:
The current behaviour returns an error when a module hook returns an error. If a hook doesn't exist, that's ignored.
For reference:

```
call_module_hook() {
        local hook="$1"
        for module in "${modules[@]}"; do
                hook_function="${module}_${hook}"
                [ "$(type -t "${hook_function}")" = "function" ] || continue # Ignore if hook doesn't exist
                "${hook_function}" && true # To not trigger errexit
                ret=$?
                [ $ret -eq 0 ] || return $ret # If call failed, return immediately
        done
        return 0
}
```

Reply:
Yes, but I changed because when calling a function that does not exist it should return an error, not ignore it, specially for the jeos-config module calls, so I just left the || continue where it was, ignoring only the hook functions non-existence and actually checking the other functions. It has the exact same effect as before.

Reply:
Time for a `call_module_optional`?

Reply:
Not sure, I spend some time trying to figure the best approach here, and still not sure, should we just guarantee that the modules have all the functions we are calling? What if we add typos by mistake (something I already did during the refactor) and we have no way to really know if the functions are being called or just being ignored because they don't really exist, as a general rule I like to treat all errors, never ignore, so we might need a little bit more code to be sure.

Reply:
Currently `call_module_hook` is used in some critical places, for hooks like `systemd_firstboot` and `post`. Currently this just makes it ignore errors.

I get your argument that `call_module` should fail if the hook isn't implemented, but for `call_module_hook` that's not ideal.

If you don't like `call_module_optional`, `call_module` could accept an optional third parameter which is returned if the hook doesn't exist.